### ISSUE TITLE
Wrong loss function in the training of encoder-decoder

### ISSUE THREAD

**Issue by Lezcano** [mindsdb/lightwood#284]
This issue comprises two different bugs.

The first one, is that the function in the encoder-decoder RNN model is not correct, as it is also adding the loss of the hidden values given by the encoder.

https://github.com/mindsdb/lightwood/blob/stable/lightwood/encoders/time_series/rnn.py#L104

On top of that, a more problematic problem and more difficult to handle is that of sequences of different length. The current implementation does not handle correctly sequences of different lengths. In particular, it assumes that the input and output are of the same length, and that every element in the dataset has the same length. At the moment, the training code does not use the `eos` value in the decoder.

When it comes to the encoder, the transformer that I will PR when I manage to run the tests handles this correctly, which could serve as reference.

More generally, I think it would be reasonable to implement the encoder-decoder in a batched way, as for larger datasets, the current implementation will prove to be too slow. At the same time, a batched implementation should be shorter and more clear to read than the current one.

As a side note, I would recommend to use the function `pad_sequence` from PyTorch, rather than using the current custom implementation to pad.

**Comment by torrmal** [comment]
@Lezcano 
You want to compute the loss on every recursion of the RNN, so I believe L104 is correct @paxcema can you give it a second pair of eyes.

As per the same length, we zero pat to support various length precisely, but also at the same time be able to deal with batches.

EOS makes sense for strings, but these sequences are not necessarily strings, @paxcema lets evaluate if it makes senses to have and end of sequence flag.

**Comment by Lezcano** [comment]
All of that is possible to do with a proper handling of sequences while keeping the batches, as shown in the `Transformer` PR.
When it comes to the EOS, this is just one token that the `RNN` will learn to output to signal that the sequence has finished. Of course, it does not make sense when considering just the `RNN` as an encoder, but at some point you will have to project that onto a categorical class to be able to know when has the decoder stopped. Once the EOS has been output, the next tokens have to be considered as garbage, and as such, they should not be included in the loss.

**Comment by Lezcano** [comment]
For reference, here is a particularly clean tutorial on how to implement an encoder/decoder with batching (padded sequences) and teacher reinforcement:

https://github.com/bentrevett/pytorch-seq2seq/blob/master/4%20-%20Packed%20Padded%20Sequences%2C%20Masking%2C%20Inference%20and%20BLEU.ipynb

In [18] you can see that the loss function is just called on the outputs of the decoder. Furthermore, the reinforcement is just done during training and in the decoder, but not the encoder, as the encoder you always assume that you will have the full sentence to translate from.

At the same time, observe the use of `sos` and `eos` in the code, and how the latter is used to determine when has a sentence actually finished during inference. At the same time, the loss during training is controlled by the length of each element, to make sure that no outputs after that length are added to the loss, to avoid fitting random noise.

**Comment by paxcema** [comment]
@Lezcano is right in that the standard way of computing the loss is just with the decoder contribution. I've tried both approaches and they seem to work similarly, but by no means has my benchmarking been thorough. I agree we can drop out the encoder contribution to the loss.

Regarding batching, the current implementation does handle batched inputs, but the code in #285 is cleaner. We should transition to that approach. 

`pad_sequence` was being considered (see #226), we can re-open that issue to simplify the RNN encoder logic.

The EOS check at decoding is a very good point, we actually do not do this. Will aim to implement it in the next release

**Comment by George3d6** [comment]
This should be closed
